### PR TITLE
Reintroduce opensearch reindexing

### PR DESCRIPTION
**Pull request by scruti**
## Review apps

https://cosmetics-pr-xxxx-submit-web.london.cloudapps.digital/
https://cosmetics-pr-xxxx-search-web.london.cloudapps.digital/

### CODE REVIEW

File: cosmetics-web/app/jobs/enqueue_reindex_opensearch_job.rb [hunk 1]
```diff
@@ -1,7 +0,0 @@
-# Because reindex job is slow, sidekiq cron is starting it several times
```
Comment (scruti) [inline cosmetics-web/app/jobs/enqueue_reindex_opensearch_job.rb:1]: No code was calling this job

File: cosmetics-web/app/jobs/reindex_opensearch_job.rb [hunk 2]
```diff
@@ -1,28 +1,47 @@
...
-    ActiveRecord::Base.transaction do
-      # tries to aquire lock. Wont block, and we need to return in case lock exist
-      lock = ActiveRecord::Base.connection.execute("SELECT pg_try_advisory_xact_lock(#{AdvisoryLock::NAMESPACES[AdvisoryLock::SEARCH]}, #{AdvisoryLock::OPERATIONS[AdvisoryLock::REINDEX_NOTIFICATIONS]})")
-      # lock.result returns such array: [[true]] or [[false]]
-      # when false, lock exists and we don't want to proceed
-      # when true, we aquired the lock and we can perform operation in current transaction
-      return unless lock.values.first.first
-
-      total = 0
+    PostgresTransactionLock.try_with_lock("reindex_notifications") do
```
Comment (slawosz) [inline cosmetics-web/app/jobs/reindex_opensearch_job.rb:15]: `PostgresDistributedLock` would be slightly better name here.

**Comment by slawosz** [comment]
One important thing: can you add tags to all log entries here, otherwise it will be very hard to follow what is happening.
So ie:
Instead of:
```
Sidekiq.logger.info "Reindexing #{model} to Opensearch #{new_index} index..."
```
it should be 
```
Sidekiq.logger.info "[Opensearch] Reindexing #{model} to Opensearch #{new_index} index..."
```

File: cosmetics-web/app/lib/postgres_transaction_lock.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Obtains exclusive Postgres transaction level advisory lock if available.
+# The lock is automatically released at the end of the current transaction and cannot be released explicitly.
+module PostgresTransactionLock
+  def self.try_with_lock(lock_name)
+    lock_id = Zlib.crc32(lock_name.to_s) # Same string will always return same id value
```
Comment (slawosz) [inline cosmetics-web/app/lib/postgres_transaction_lock.rb:5]: Really good find!

File: cosmetics-web/app/jobs/reindex_opensearch_job.rb (hunk 2 above)
Comment (slawosz) [inline cosmetics-web/app/jobs/reindex_opensearch_job.rb:15]: Oh,
and I would keep names of all locks in constants, as kind of good practice, so something like `AdvisoryLock::REINDEX_NOTIFICATIONS` or so here.

**Comment by scruti** [comment]
Thanks for the review @slawosz , regarding tagging the Sidekiq log entries, the Sidekiq logger automatically includes the job class name in its log entries, so manual tagging in the message would be redundant.

EG of the output with the current code:
```
2022-12-13T16:38:36.240Z 14 TID-1te2 ReindexOpensearchJob JID-688c4710bf6a35a45af00622 INFO: Swapped Opensearch index for Notification from cosmetics-pr-2706_production_notifications_20221212181910 to cosmetics-pr-2706_production_notifications_20221213162507
```

**Review by slawosz — APPROVED** [review]

**Review by slawosz — CHANGES_REQUESTED** [review]
I need to do more testing before approving

**Review by slawosz — APPROVED** [review]
